docs/architecture/adr-110-e2e-abci++.md

```diff
@@ -0,0 +1,267 @@
+# ADR 82: E2E tests for CometBFT's behaviour in respect to ABCI++.
+
+## Context
+
+We want to test whether CommetBFT respects the ABCI++ grammar. To do this, we need to enhance the e2e tests infrastructure. Specifically, we plan to do three things:
+- Log every CometBFT's ABCI++ request during the execution.
+- Parse the logs post-mortem and extract all ABCI++ requests.
+- Check if the set of observed requests respects the ABCI++ grammar.
+
+
+Issue: [353](https://github.com/cometbft/cometbft/issues/353).
+
+Current version does not support vote extensions. However, this is the next step. 
+
+
+## Decision
+
+### 1) ABCI++ requests logging
+The idea was to do this at the Application side. Every time the Application 
+receives a request, it logs it.
+
+<strong>Implementation</strong>
+
+The rationale behind this part of the implementation was to log the request concisely and use the existing structures as much as possible. 
+
+Whenever an ABCI request is made, the application will create `abci.Request` (`abci` stands for `"github.com/cometbft/cometbft/abci/types"`) and log it.  The example is below.  
+
+```go
+func (app *Application) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+
+	r := &abci.Request{Value: &abci.Request_InitChain{InitChain: &abci.RequestInitChain{}}}
+	app.logRequest(r)
+    
+	...
+}
+```
+Notice here that we create an empty `abci.RequestInitChain` object while we can also use the one passed to the `InitChain` function. The reason behind this is that, at the moment, we do not need specific fields of the request; we just need to be able to extract the information about the request type. For this, an empty object of a particular type is enough. 
+
+`app.logRequest(r)` function is a new function implemented in the same file (`test/e2e/app/app.go`). Its implementation is the following: 
+
+```go
+func (app *Application) logRequest(req *abci.Request) {
+	s, err := GetABCIRequestString(req)
+	if err != nil {
+		panic(err)
+	}
+	app.logger.Debug(s)
```

Comment:
Debug is an option, but, reflecting on it, wouldn't it be more convenient to turn them into "info", and have a configuration option in the e2e app that enables disables these logs? (it would just mean an extra if here).

The problem I see with debug is that the debug output is unbearable ATM if we enable it for all modules.

Reply:
This could also be the same bool config to enable/disable this feature when running e2e tests

Reply:
I have no specific reasons why I am using `Debug` here, I think we discussed it a long time ago, and we considered it to be the `Debug`.  Anyway, I will do what you suggest here.